### PR TITLE
Fix `on_train_batch_start` hook to end epoch early

### LINKED ISSUE #3693: Missing attribute "training_step_output_for_epoch_end"
I used the documentation way of stopping the training (https://pytorch-lightning.readthedocs.io/en/latest/early_stopping.html#enable-early-stopping-using-callbacks-on-epoch-end). 

If on_bath_start method returns -1 at the very beginning of an epoch, the titled AttributeError exception.
The problem is in training_loop.py line 496 (batch_output.training_step_output_for_epoch_end).

#### Code sample
Use the method and run your code:
```python
    def on_batch_start(self, batch):
        return -1
```

### Expected behavior
Check batch_output value if equals -1 before running trainin_loop.py line 495. 
The early stopping method achieved the same way the documentation specifies should not throw an exception but rather simply stop the training.

### Environment
* CUDA:
	- GPU:
	- available:         False
	- version:           None
* Packages:
	- numpy:             1.19.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0
	- pytorch-lightning: 0.9.0
	- tqdm:              4.49.0
* System:
	- OS:                Windows
	- architecture:
		- 64bit
		- WindowsPE
	- processor:         Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
	- python:            3.8.5
	- version:           10.0.18362

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Fixes #3693 

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ydcjeff** [comment]
It's ready to review.

File: tests/models/test_hooks.py [hunk 1]
```diff
@@ -108,3 +108,28 @@ def transfer_batch_to_device(self, data, device):
...
+
+@pytest.mark.parametrize(
+    ['max_epochs', 'batch_idx_'],
+    [
+        pytest.param(2, 5),
+        pytest.param(3, 8),
+        pytest.param(4, 12)
+    ]
+)
+def test_on_train_batch_start_hook(max_epochs, batch_idx_):
```
Comment (Borda) [inline tests/models/test_hooks.py:121]:
```suggestion
@pytest.mark.parametrize(
    'max_epochs,batch_idx_',
    [(2, 5), (3, 8), (4, 12)]
)
def test_on_train_batch_start_hook(max_epochs, batch_idx_):
```

**Review by Borda — APPROVED** [review]
lgtm

**Review by SkafteNicki — APPROVED** [review]
Lgtm

**Review by rohitgr7 — APPROVED** [review]
LGTM :)

**Pull request merged**